File: apps/server/aetherblog-service/blog-service/src/main/java/com/aetherblog/blog/service/FolderService.java
package com.aetherblog.blog.service;

import com.aetherblog.blog.entity.MediaFolder;

import java.util.List;

/**
 * 文件夹服务接口
 *
 * @ref 媒体库深度优化方案 - Phase 1: 文件夹层级管理
 * @author AI Assistant
 * @since 2.1.0
 */
public interface FolderService {

    /**
     * 创建文件夹
     *
     * @param name 文件夹名称
     * @param description 描述
     * @param parentId 父文件夹ID (null表示根目录)
     * @param userId 创建者ID
     * @return 创建的文件夹
     */
    MediaFolder create(String name, String description, Long parentId, Long userId);

    /**
     * 更新文件夹
     *
     * @param id 文件夹ID
     * @param name 新名称
     * @param description 新描述
     * @param color 颜色
     * @param icon 图标
     * @param userId 更新者ID
     * @return 更新后的文件夹
     */
    MediaFolder update(Long id, String name, String description, String color, String icon, Long userId);

    /**
     * 删除文件夹 (级联删除子文件夹和文件)
     *
     * @param id 文件夹ID
     */
    void delete(Long id);

    /**
     * 根据ID获取文件夹
     *
     * @param id 文件夹ID
     * @return 文件夹实体
     */
    MediaFolder getById(Long id);

    /**
     * 根据路径获取文件夹
     *
     * @param path 文件夹路径
     * @return 文件夹实体
     */
    MediaFolder getByPath(String path);

    /**
     * 获取完整文件夹树
     *
     * @return 文件夹树列表
     */
    List<MediaFolder> getTree();

    /**
     * 获取指定用户可见的文件夹树
     *
     * @param userId 用户ID
     * @return 文件夹树列表
     */
    List<MediaFolder> getTreeByUserId(Long userId);

    /**
     * 获取父文件夹的直接子文件夹
     *
     * @param parentId 父文件夹ID (null表示根目录)
     * @return 子文件夹列表
     */
    List<MediaFolder> getChildren(Long parentId);

    /**
     * 移动文件夹到新的父文件夹
     *
     * @param folderId 要移动的文件夹ID
     * @param newParentId 新父文件夹ID (null表示移动到根目录)
     * @param userId 操作者ID
     * @return 移动后的文件夹
     */
    MediaFolder move(Long folderId, Long newParentId, Long userId);

    /**
     * 更新文件夹统计信息 (文件数量和总大小)
     *
     * @param folderId 文件夹ID
     */
    void updateStatistics(Long folderId);

    /**
     * 递归更新文件夹及其所有父文件夹的统计信息
     *
     * @param folderId 文件夹ID
     */
    void updateStatisticsRecursive(Long folderId);

    /**
     * 检查用户是否有权限访问文件夹
     *
     * @param folderId 文件夹ID
     * @param userId 用户ID
     * @return 如果用户拥有访问权限则返回 true
     */
    boolean hasAccess(Long folderId, Long userId);

    /**
     * 生成唯一的slug
     *
     * @param name 文件夹名称
     * @param parentId 父文件夹ID
     * @return 唯一的slug
     */
    String generateUniqueSlug(String name, Long parentId);
}

File: apps/server-go/internal/service/folder_service.go
package service

import (
	"context"

	"github.com/golovin0623/aetherblog-server/internal/dto"
	"github.com/golovin0623/aetherblog-server/internal/model"
	"github.com/golovin0623/aetherblog-server/internal/repository"
)

// FolderService 管理媒体文件夹的业务逻辑。
type FolderService struct {
	repo *repository.FolderRepo
}

// NewFolderService 使用给定的仓储创建 FolderService 实例。
func NewFolderService(repo *repository.FolderRepo) *FolderService {
	return &FolderService{repo: repo}
}

// GetTree 返回所有文件夹，并组装为父子嵌套树形结构。
func (s *FolderService) GetTree(ctx context.Context) ([]dto.MediaFolderVO, error) {
	folders, err := s.repo.FindAll(ctx)
	if err != nil {
		return nil, err
	}
	return buildFolderTree(toFolderVOs(folders)), nil
}

// GetByID 通过主键查询单个文件夹，不存在时返回 nil, nil。
func (s *FolderService) GetByID(ctx context.Context, id int64) (*dto.MediaFolderVO, error) {
	f, err := s.repo.FindByID(ctx, id)
	if err != nil || f == nil {
		return nil, err
	}
	vo := toFolderVO(*f)
	return &vo, nil
}

// GetOwnerID 返回指定文件夹的 owner_id，用于 handler 层 ownership 校验。
// 文件夹不存在时返回 (nil, nil)。
func (s *FolderService) GetOwnerID(ctx context.Context, id int64) (*int64, error) {
	f, err := s.repo.FindByID(ctx, id)
	if err != nil || f == nil {
		return nil, err
	}
	return f.OwnerID, nil
}

// GetChildren 返回指定文件夹的直接子文件夹列表（平铺，不递归）。
func (s *FolderService) GetChildren(ctx context.Context, id int64) ([]dto.MediaFolderVO, error) {
	fs, err := s.repo.FindChildren(ctx, id)
	if err != nil {
		return nil, err
	}
	return toFolderVOs(fs), nil
}

// EnsureFolderByPath 是 KB 等内部模块用的目录确保入口。详见 FolderRepo.EnsureFolderByPath。
// 返回叶子目录的 VO；segments 不包含 "root"。
func (s *FolderService) EnsureFolderByPath(ctx context.Context, segments []string, ownerID *int64, isSystem, undeletable bool) (*dto.MediaFolderVO, error) {
	leaf, err := s.repo.EnsureFolderByPath(ctx, segments, ownerID, isSystem, undeletable)
	if err != nil {
		return nil, err
	}
	if leaf == nil {
		return nil, nil
	}
	vo := toFolderVO(*leaf)
	return &vo, nil
}

// FindByPath 按物化路径查询单个目录（含系统目录）。
func (s *FolderService) FindByPath(ctx context.Context, path string) (*dto.MediaFolderVO, error) {
	f, err := s.repo.FindByPath(ctx, path)
	if err != nil || f == nil {
		return nil, err
	}
	vo := toFolderVO(*f)
	return &vo, nil
}

// Create 创建新文件夹。
// 业务规则：未指定可见性时默认设为 PRIVATE。
func (s *FolderService) Create(ctx context.Context, req dto.FolderRequest, ownerID *int64) (*dto.MediaFolderVO, error) {
	vis := req.Visibility
	if vis == "" {
		vis = "PRIVATE" // 可见性默认为私有
	}
	f, err := s.repo.Create(ctx, repository.FolderRequest{
		Name:        req.Name,
		Description: req.Description,
		ParentID:    req.ParentID,
		Color:       req.Color,
		Icon:        req.Icon,
		Visibility:  vis,
		OwnerID:     ownerID,
	})
	if err != nil {
		return nil, err
	}
	vo := toFolderVO(*f)
	return &vo, nil
}

// Update 修改文件夹的展示属性（名称、描述、颜色、图标、可见性）。
// 未指定可见性时默认设为 PRIVATE。
func (s *FolderService) Update(ctx context.Context, id int64, req dto.FolderRequest, ownerID *int64) error {
	vis := req.Visibility
	if vis == "" {
		vis = "PRIVATE"
	}
	return s.repo.Update(ctx, id, repository.FolderRequest{
		Name:        req.Name,
		Description: req.Description,
		Color:       req.Color,
		Icon:        req.Icon,
		Visibility:  vis,
		OwnerID:     ownerID,
	})
}

// Delete 永久删除指定文件夹。
func (s *FolderService) Delete(ctx context.Context, id int64) error {
	return s.repo.Delete(ctx, id)
}

// Move 将文件夹重新挂载到 newParentID 下；newParentID 为 nil 表示移至根级别。
func (s *FolderService) Move(ctx context.Context, id int64, newParentID *int64, updatedBy *int64) error {
	return s.repo.Move(ctx, id, newParentID, updatedBy)
}

// --- 内部辅助函数 ---

// toFolderVO 将单个 model.MediaFolder 转换为 dto.MediaFolderVO。
func toFolderVO(f model.MediaFolder) dto.MediaFolderVO {
	return dto.MediaFolderVO{
		ID:          f.ID,
		Name:        f.Name,
		Slug:        f.Slug,
		Description: f.Description,
		ParentID:    f.ParentID,
		Path:        f.Path,
		Depth:       f.Depth,
		Color:       f.Color,
		Icon:        f.Icon,
		Visibility:  f.Visibility,
		FileCount:   f.FileCount,
		TotalSize:   f.TotalSize,
		IsSystem:    f.IsSystem,
		Undeletable: f.Undeletable,
		CreatedAt:   f.CreatedAt,
	}
}

// toFolderVOs 将 model.MediaFolder 切片批量转换为 dto.MediaFolderVO 切片。
func toFolderVOs(fs []model.MediaFolder) []dto.MediaFolderVO {
	vos := make([]dto.MediaFolderVO, len(fs))
	for i, f := range fs {
		vos[i] = toFolderVO(f)
	}
	return vos
}

// buildFolderTree 将平铺的文件夹 VO 列表按 parent_id 构建为嵌套树形结构。
// 使用 O(N) 一次遍历哈希表分组，然后自顶向下递归组装。
// 原两轮指针挂载方案在值拷贝场景下会丢失孙级及更深层级的嵌套数据（先挂载后拷贝的根节点
// 不含后续才挂上的子节点），此处统一为"先分组、后递归"的安全模式修复。
// ref: PR#713 —— 修复深层嵌套丢失 + O(N) 优化 + 为 map 预分配容量。
func buildFolderTree(vos []dto.MediaFolderVO) []dto.MediaFolderVO {
	if len(vos) == 0 {
		return nil
	}

	var roots []dto.MediaFolderVO
	byParent := make(map[int64][]dto.MediaFolderVO, len(vos))

	for i := range vos {
		if vos[i].ParentID == nil {
			roots = append(roots, vos[i])
		} else {
			pid := *vos[i].ParentID
			byParent[pid] = append(byParent[pid], vos[i])
		}
	}

	var build func(nodes []dto.MediaFolderVO) []dto.MediaFolderVO
	build = func(nodes []dto.MediaFolderVO) []dto.MediaFolderVO {
		for i := range nodes {
			if children, ok := byParent[nodes[i].ID]; ok {
				nodes[i].Children = build(children)
			}
		}
		return nodes
	}

	return build(roots)
}
